### DISCUSSION TITLE
The editor changes history\watcher

### DISCUSSION THREAD

**Discussion by Iipal** [ueberdosis/tiptap#2459] [General Questions]
Hi, I need somehow to get the information only about changes in the editor. How can I do this? Was looking for about 2 hours on changing between objects on this events:

```js
    onUpdate: (opts) => {
      console.log("onUpdate", opts, JSON.stringify(opts.editor.getJSON(), null, "  "));
    },
    onTransaction: (opts) => {
      console.log("onTransaction", opts);
    },
    onSelectionUpdate: (opts) => {
      console.log("onSelectionUpdate", opts);
    },
```

and can't really tell where is stored what I looking for. I mean changes like: an index of Node(parapgraph\heading\customExtensionComponent) where was added changes, position start-end and so on....

Thanks in advance.

**Comment by philippkuehn** [comment]
This is a hard topic and can be solved in different ways. There is a simple `getChangedRanges` helper that return the changed ranges of a transform/transaction.

Example:

https://github.com/ueberdosis/tiptap/blob/5144a6cd1b12a161b5879efab22d2dda0c2a20ba/packages/extension-link/src/helpers/autolink.ts#L30


A more advanced solution is to use [prosemirror-changeset](https://github.com/ProseMirror/prosemirror-changeset).
